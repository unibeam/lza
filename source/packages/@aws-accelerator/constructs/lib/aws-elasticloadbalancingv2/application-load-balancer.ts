/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { pascalCase } from 'change-case';
import { ApplicationLoadBalancerListenerConfig } from '@aws-accelerator/config';

export interface IApplicationLoadBalancerResource extends cdk.IResource {
  /**
   * The ARN of the ApplicationLoadBalancer.
   */
  readonly applicationLoadBalancerArn: string;
  /**
   * The name of the ApplicationLoadBalancer
   */
  readonly applicationLoadBalancerName: string;
}

export type albAttributesType = {
  deletionProtection?: boolean;
  idleTimeout?: number;
  routingHttpDesyncMitigationMode?: string;
  routingHttpDropInvalidHeader?: boolean;
  routingHttpXAmznTlsCipherEnable?: boolean;
  routingHttpXffClientPort?: boolean;
  routingHttpXffHeaderProcessingMode?: 'append' | 'preserve' | 'remove';
  http2Enabled?: boolean;
  wafFailOpen?: boolean;
};

export type albForwardTargetGroup = {
  targetGroupArn: string;
};

export type albListenerActionProperty = {
  type: string;
  fixedResponseConfig?: {
    statusCode: string;

    // the properties below are optional
    contentType?: string;
    messageBody?: string;
  };
  forwardConfig?: {
    targetGroups: albForwardTargetGroup[];
    targetGroupStickinessConfig?: {
      durationSeconds?: number;
      enabled?: boolean;
    };
  };
  order?: number;
  redirectConfig?: {
    statusCode?: string;

    // the properties below are optional
    host?: string;
    path?: string;
    port?: string;
    protocol?: string;
    query?: string;
  };
  targetGroupArn: string;
};

export interface ApplicationLoadBalancerProps {
  /**
   * Name for Application Load Balancer.
   */
  readonly name: string;
  /**
   * SSM parameter prefix
   */
  readonly ssmPrefix: string;
  /**
   * Application Load Balancer Subnets (required).
   */
  readonly subnets: string[];
  /**
   * Application Load Balancer SecurityGroups (required).
   */
  readonly securityGroups?: string[];
  /**
   * Application Load Balancer scheme.
   */
  readonly scheme?: string;
  /**
   * Application Load Balancer attributes.
   */
  readonly attributes?: albAttributesType;
  /**
   * Listeners for Application Load Balancer.
   */
  readonly listeners?: ApplicationLoadBalancerListenerConfig[];

  /**
   * Access logs s3 bucket name.
   */
  readonly accessLogsBucket: string;
}

export class ApplicationLoadBalancer extends cdk.Resource implements IApplicationLoadBalancerResource {
  public readonly applicationLoadBalancerArn: string;
  public readonly applicationLoadBalancerName: string;
  constructor(scope: Construct, id: string, props: ApplicationLoadBalancerProps) {
    super(scope, id);
    const resource = new cdk.aws_elasticloadbalancingv2.CfnLoadBalancer(this, 'Resource', {
      type: 'application',
      subnets: props.subnets,
      name: props.name,
      scheme: props.scheme,
      securityGroups: props.securityGroups,
      loadBalancerAttributes: this.buildAttributes(props),
    });

    // Add Name tag
    cdk.Tags.of(this).add('Name', props.name);

    // Set initial properties
    this.applicationLoadBalancerArn = resource.ref;
    this.applicationLoadBalancerName = resource.attrLoadBalancerName;

    for (const listener of props.listeners ?? []) {
      const listenerAction: cdk.aws_elasticloadbalancingv2.CfnListener.ActionProperty =
        this.getListenerAction(listener);
      new cdk.aws_elasticloadbalancingv2.CfnListener(this, pascalCase(`Listener${listener.name}`), {
        defaultActions: [listenerAction],
        loadBalancerArn: resource.ref,
        certificates: [{ certificateArn: this.getCertificate(props.ssmPrefix, listener.certificate) }],
        port: listener.port,
        protocol: listener.protocol,
        sslPolicy: listener.sslPolicy!,
      });
    }
  }
  private getCertificate(ssmPrefix: string, certificate: string | undefined) {
    if (certificate) {
      //check if user provided arn. If so do nothing, if not get it from ssm
      if (certificate.match('\\arn:*')) {
        return certificate;
      } else {
        return cdk.aws_ssm.StringParameter.valueForStringParameter(this, `${ssmPrefix}/acm/${certificate}/arn`);
      }
    }
    return undefined;
  }
  private getListenerAction(
    listener: ApplicationLoadBalancerListenerConfig,
  ): cdk.aws_elasticloadbalancingv2.CfnListener.ActionProperty {
    const listenerTargetGroupArn = listener.targetGroup;
    const actionValues: albListenerActionProperty = {
      type: listener.type,
      order: listener.order,
      targetGroupArn: listenerTargetGroupArn,
    };
    switch (listener.type) {
      case 'forward':
        actionValues.forwardConfig = {
          targetGroups: [{ targetGroupArn: listener.targetGroup }],
          targetGroupStickinessConfig: listener.forwardConfig?.targetGroupStickinessConfig ?? undefined,
        };
        break;
      case 'redirect':
        if (listener.redirectConfig) {
          actionValues.redirectConfig = {
            host: listener.redirectConfig.host,
            path: listener.redirectConfig.path,
            port: listener.redirectConfig.port?.toString() ?? undefined,
            protocol: listener.redirectConfig.protocol,
            query: listener.redirectConfig.query,
            statusCode: listener.redirectConfig.statusCode,
          };
        } else {
          throw new Error(`Listener ${listener.name} is set to redirect but redirectConfig is not defined`);
        }
        break;
      case 'fixed-response':
        if (listener.fixedResponseConfig) {
          actionValues.fixedResponseConfig = {
            contentType: listener.fixedResponseConfig.contentType,
            messageBody: listener.fixedResponseConfig.messageBody,
            statusCode: listener.fixedResponseConfig.statusCode,
          };
        } else {
          throw new Error(`Listener ${listener.name} is set to fixed-response but fixedResponseConfig is not defined`);
        }
        break;
      default:
        throw new Error(`Listener ${listener.name} is not set to forward, fixed-response or redirect`);
    }

    return actionValues as cdk.aws_elasticloadbalancingv2.CfnListener.ActionProperty;
  }

  private buildAttributes(props: ApplicationLoadBalancerProps) {
    // add elements to the array.
    // based on https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-elasticloadbalancingv2-loadbalancer-loadbalancerattributes.html
    const albAttributesProperties = [
      {
        key: 'access_logs.s3.enabled',
        value: 'true',
      },
      {
        key: 'access_logs.s3.bucket',
        value: props.accessLogsBucket,
      },
      {
        key: 'access_logs.s3.prefix',
        value: `${cdk.Stack.of(this).account}/${cdk.Stack.of(this).region}/${props.name}`,
      },
    ];
    if (!props.attributes) {
      return albAttributesProperties;
    }

    if (props.attributes.deletionProtection) {
      albAttributesProperties.push({
        key: 'deletion_protection.enabled',
        value: props.attributes.deletionProtection.toString(),
      });
    }
    if (props.attributes.idleTimeout) {
      albAttributesProperties.push({
        key: 'idle_timeout.timeout_seconds',
        value: props.attributes.idleTimeout.toString(),
      });
    }
    if (props.attributes.routingHttpDesyncMitigationMode) {
      albAttributesProperties.push({
        key: 'routing.http.desync_mitigation_mode',
        value: props.attributes.routingHttpDesyncMitigationMode,
      });
    }
    if (props.attributes.routingHttpDropInvalidHeader) {
      albAttributesProperties.push({
        key: 'routing.http.drop_invalid_header_fields.enabled',
        value: props.attributes.routingHttpDropInvalidHeader.toString(),
      });
    }
    if (props.attributes.routingHttpXAmznTlsCipherEnable) {
      albAttributesProperties.push({
        key: 'routing.http.x_amzn_tls_version_and_cipher_suite.enabled',
        value: props.attributes.routingHttpXAmznTlsCipherEnable.toString(),
      });
    }
    if (props.attributes.routingHttpXffClientPort) {
      albAttributesProperties.push({
        key: 'routing.http.xff_client_port.enabled',
        value: props.attributes.routingHttpXffClientPort.toString(),
      });
    }
    if (props.attributes.routingHttpXffHeaderProcessingMode) {
      albAttributesProperties.push({
        key: 'routing.http.xff_header_processing.mode',
        value: props.attributes.routingHttpXffHeaderProcessingMode,
      });
    }
    if (props.attributes.http2Enabled) {
      albAttributesProperties.push({
        key: 'routing.http2.enabled',
        value: props.attributes.http2Enabled.toString(),
      });
    }
    if (props.attributes.wafFailOpen) {
      albAttributesProperties.push({
        key: 'waf.fail_open.enabled',
        value: props.attributes.wafFailOpen.toString(),
      });
    }

    return albAttributesProperties;
  }
}
