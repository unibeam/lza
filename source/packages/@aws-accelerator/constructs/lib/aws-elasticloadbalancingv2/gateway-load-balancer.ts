/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface IGatewayLoadBalancer extends cdk.IResource {
  /**
   * The ID of the VPC endpoint service associated with this Gateway Load Balancer.
   */
  readonly endpointServiceId: string;
  /**
   * The ARN of the Gateway Load Balancer.
   */
  readonly loadBalancerArn: string;
  /**
   * The name of the Gateway Load Balancer
   */
  readonly loadBalancerName: string;
}

export interface GatewayLoadBalancerProps {
  /**
   * The name of the Gateway Load Balancer.
   */
  readonly name: string;
  /**
   * An array of account principals allowed to create endpoints for the service.
   */
  readonly allowedPrincipals: string[];
  /**
   * The subnets the Gateway Load Balancer will be deployed to.
   */
  readonly subnets: string[];
  /**
   * Whether to enable cross-zone load balancing.
   *
   * @default 'true'
   */
  readonly crossZoneLoadBalancing?: boolean;
  /**
   * Whether to enable deletion protection.
   *
   * @default 'false'
   */
  readonly deletionProtection?: boolean;
  /**
   * An array of CloudFormation tags to apply to the Gateway Load Balancer.
   */
  readonly tags?: cdk.CfnTag[];
}

export class GatewayLoadBalancer extends cdk.Resource implements IGatewayLoadBalancer {
  public readonly endpointServiceId: string;
  public readonly loadBalancerArn: string;
  public readonly loadBalancerName: string;
  private allowedPrincipals: string[];
  private endpointService: cdk.aws_ec2.CfnVPCEndpointService;

  constructor(scope: Construct, id: string, props: GatewayLoadBalancerProps) {
    super(scope, id);

    const resource = new cdk.aws_elasticloadbalancingv2.CfnLoadBalancer(this, 'Resource', {
      loadBalancerAttributes: [
        {
          key: 'deletion_protection.enabled',
          value: props.deletionProtection ? props.deletionProtection.toString() : 'false',
        },
        {
          key: 'load_balancing.cross_zone.enabled',
          value: props.crossZoneLoadBalancing ? props.crossZoneLoadBalancing.toString() : 'true',
        },
      ],
      subnets: props.subnets,
      tags: props.tags,
      type: 'gateway',
    });
    // Add Name tag
    cdk.Tags.of(this).add('Name', props.name);

    // Set initial properties
    this.loadBalancerArn = resource.ref;
    this.loadBalancerName = resource.attrLoadBalancerName;

    // Create endpoint service
    this.endpointService = this.createEndpointService();
    this.allowedPrincipals = props.allowedPrincipals;
    this.endpointServiceId = this.endpointService.ref;
    if (this.allowedPrincipals.length > 0) {
      this.createEndpointServicePermissions();
    }
  }

  /**
   * Create endpoint service for the load balancer.
   */
  private createEndpointService(): cdk.aws_ec2.CfnVPCEndpointService {
    return new cdk.aws_ec2.CfnVPCEndpointService(this, 'EndpointService', {
      acceptanceRequired: false,
      gatewayLoadBalancerArns: [this.loadBalancerArn],
    });
  }

  /**
   * Create endpoint service permissions.
   */
  private createEndpointServicePermissions(): void {
    const principals = this.allowedPrincipals.map(item => {
      return `arn:${cdk.Aws.PARTITION}:iam::${item}:root`;
    });

    new cdk.aws_ec2.CfnVPCEndpointServicePermissions(this, 'EndpointServicePermissions', {
      serviceId: this.endpointServiceId,
      allowedPrincipals: principals,
    });
  }
}
