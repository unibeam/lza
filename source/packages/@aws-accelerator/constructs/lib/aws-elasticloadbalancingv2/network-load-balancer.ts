/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { pascalCase } from 'change-case';

export interface INetworkLoadBalancerResource extends cdk.IResource {
  /**
   * The ARN of the NetworkLoadBalancer.
   */
  readonly networkLoadBalancerArn: string;
  /**
   * The name of the NetworkLoadBalancer
   */
  readonly networkLoadBalancerName: string;
}

export type NetworkLoadBalancerListener = {
  /**
   * Name for Listener.
   */
  readonly name: string;
  /**
   * ACM ARN of the certificate to be associated with the listener.
   */
  readonly certificate?: string;
  /**
   * Port where the traffic is directed to.
   */
  readonly port?: number;
  /**
   * Protocol used for the traffic.
   */
  readonly protocol?: string;
  /**
   * Application-Layer Protocol Negotiation (ALPN) policy for TLS encrypted traffic
   */
  readonly alpnPolicy?: string;
  /**
   * SSL policy for TLS encrypted traffic
   */
  readonly sslPolicy?: string;
  /**
   * Target Group to direct the traffic to.
   */
  readonly targetGroup: string;
};
export interface NetworkLoadBalancerProps {
  /**
   * Name for Network Load Balancer.
   */
  readonly name: string;
  /**
   * SSM prefix
   */
  readonly ssmPrefix: string;
  /**
   * Network Load Balancer Subnets (required).
   */
  readonly subnets: string[];
  /**
   * Network Load Balancer scheme.
   */
  readonly scheme?: string;
  /**
   * Network Load Balancer deletionProtection
   */
  readonly deletionProtection?: boolean;
  /**
   * Cross Zone load balancing for Network Load Balancer.
   */
  readonly crossZoneLoadBalancing?: boolean;
  /**
   * Listeners for Network Load Balancer.
   */
  readonly listeners?: NetworkLoadBalancerListener[];
  /**
   * Access logs s3 bucket name.
   */
  readonly accessLogsBucket: string;
  /**
   * VPC Name (required).
   */
  readonly vpcName: string;
  /**
   *  App Name (required).
   */
  readonly appName: string;
}

export class NetworkLoadBalancer extends cdk.Resource implements INetworkLoadBalancerResource {
  public readonly networkLoadBalancerArn: string;
  public readonly networkLoadBalancerName: string;
  constructor(scope: Construct, id: string, props: NetworkLoadBalancerProps) {
    super(scope, id);
    const resource = new cdk.aws_elasticloadbalancingv2.CfnLoadBalancer(this, 'Resource', {
      type: 'network',
      subnets: props.subnets,
      name: props.name,
      scheme: props.scheme,
      loadBalancerAttributes: [
        {
          key: 'deletion_protection.enabled',
          value: props.deletionProtection ? props.deletionProtection.toString() : 'false',
        },
        {
          key: 'load_balancing.cross_zone.enabled',
          value: props.crossZoneLoadBalancing ? props.crossZoneLoadBalancing.toString() : 'true',
        },
        {
          key: 'access_logs.s3.enabled',
          value: 'true',
        },
        {
          key: 'access_logs.s3.bucket',
          value: props.accessLogsBucket,
        },
        {
          key: 'access_logs.s3.prefix',
          value: `${cdk.Stack.of(this).account}/${cdk.Stack.of(this).region}/${props.name}`,
        },
      ],
    });

    // Add Name tag
    cdk.Tags.of(this).add('Name', props.name);

    // Set initial properties
    this.networkLoadBalancerArn = resource.ref;
    this.networkLoadBalancerName = resource.attrLoadBalancerName;

    for (const listener of props.listeners ?? []) {
      const targetGroupArn = this.getTargetGroupArn(
        listener.targetGroup,
        props.vpcName,
        props.appName,
        props.ssmPrefix,
      );
      new cdk.aws_elasticloadbalancingv2.CfnListener(this, pascalCase(`Listener${listener.name}`), {
        defaultActions: [
          {
            type: 'forward',
            forwardConfig: {
              targetGroups: [
                {
                  targetGroupArn: targetGroupArn,
                },
              ],
            },
            targetGroupArn: targetGroupArn,
          },
        ],
        loadBalancerArn: resource.ref,
        alpnPolicy: [listener.alpnPolicy!],
        certificates: [{ certificateArn: listener.certificate! }],
        port: listener.port!,
        protocol: listener.protocol!,
        sslPolicy: listener.sslPolicy!,
      });
    }
  }
  private getTargetGroupArn(targetGroup: string, vpcName: string, appName: string, ssmPrefix: string): string {
    if (targetGroup.match('\\${ACCEL_LOOKUP::TargetGroup:([a-zA-Z0-9-/:]*)}')) {
      const targetGroupMatch = targetGroup.match('\\${ACCEL_LOOKUP::TargetGroup:([a-zA-Z0-9-/:]*)}');
      const targetGroupValue = cdk.aws_ssm.StringParameter.valueForStringParameter(
        this,
        `${ssmPrefix}/application/targetGroup/${appName}/${vpcName}/${targetGroupMatch![1]}/arn`,
      );
      return targetGroupValue;
    } else if (targetGroup.match('\\arn:*')) {
      return targetGroup;
    } else {
      return cdk.aws_ssm.StringParameter.valueForStringParameter(
        this,
        `${ssmPrefix}/application/targetGroup/${appName}/${vpcName}/${targetGroup}/arn`,
      );
    }
  }
}
