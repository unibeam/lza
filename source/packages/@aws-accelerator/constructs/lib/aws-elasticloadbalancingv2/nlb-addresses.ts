/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { NagSuppressions } from 'cdk-nag';
import { NlbTargetTypeConfig } from '@aws-accelerator/config';

const path = require('path');

export interface INLBAddresses extends cdk.IResource {
  /**
   * The IP addresses of the endpoint.
   */
  readonly ipAddresses: cdk.Reference;
}

export interface NLBAddressesProps {
  /**
   * The ip and NLB targets
   */
  readonly targets: (NlbTargetTypeConfig | string)[];
  /**
   * The role to assume to retrieve the ip addresses
   */
  readonly assumeRoleName: string;
  /**
   * The partition
   */
  readonly partition: string;
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly kmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays: number;
}

/**
 * Class for FMSOrganizationAdminAccount
 */
export class NLBAddresses extends cdk.Resource implements INLBAddresses {
  public readonly ipAddresses: cdk.Reference;
  constructor(scope: Construct, id: string, props: NLBAddressesProps) {
    super(scope, id);
    const functionId = `${id}ProviderLambda`;
    const providerId = `${id}Provider`;

    const providerLambda = new cdk.aws_lambda.Function(this, functionId, {
      code: cdk.aws_lambda.Code.fromAsset(path.join(__dirname, 'nlb-ip-lookup/dist')),
      runtime: cdk.aws_lambda.Runtime.NODEJS_18_X,
      timeout: cdk.Duration.seconds(15),
      handler: 'index.handler',
    });

    providerLambda.addToRolePolicy(
      new cdk.aws_iam.PolicyStatement({
        sid: 'StsAssumeRole',
        effect: cdk.aws_iam.Effect.ALLOW,
        actions: ['sts:AssumeRole'],
        resources: [`arn:${cdk.Stack.of(this).partition}:iam::*:role/${props.assumeRoleName}`],
      }),
    );

    const provider = new cdk.custom_resources.Provider(this, providerId, {
      onEventHandler: providerLambda,
    });

    const logGroup = new cdk.aws_logs.LogGroup(this, `${providerLambda.node.id}LogGroup`, {
      logGroupName: `/aws/lambda/${providerLambda.functionName}`,
      retention: props.logRetentionInDays,
      encryptionKey: props.kmsKey,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    const resource = new cdk.CustomResource(this, `Resource`, {
      serviceToken: provider.serviceToken,
      properties: {
        region: cdk.Stack.of(this).region,
        targets: props.targets,
        assumeRoleName: props.assumeRoleName,
        partition: cdk.Stack.of(scope).partition,
      },
    });
    // Ensure that the LogGroup is created by Cloudformation prior to Lambda execution
    resource.node.addDependency(logGroup);
    this.ipAddresses = resource.getAtt('ipAddresses');

    const stack = cdk.Stack.of(scope);

    // AwsSolutions-IAM4: The IAM user, role, or group uses AWS managed policies
    NagSuppressions.addResourceSuppressionsByPath(
      stack,
      `${stack.stackName}/${id}/${functionId}/ServiceRole/Resource`,
      [
        {
          id: 'AwsSolutions-IAM4',
          reason: 'AWS Custom resource provider framework-role created by cdk.',
        },
      ],
    );

    // AwsSolutions-IAM5: The IAM entity contains wildcard permissions
    NagSuppressions.addResourceSuppressionsByPath(
      stack,
      `${stack.stackName}/${id}/${functionId}/ServiceRole/DefaultPolicy/Resource`,
      [
        {
          id: 'AwsSolutions-IAM5',
          reason: 'Allows only specific policy.',
        },
      ],
    );

    // AwsSolutions-IAM4: The IAM user, role, or group uses AWS managed policies
    NagSuppressions.addResourceSuppressionsByPath(
      stack,
      `${stack.stackName}/${id}/${providerId}/framework-onEvent/ServiceRole/Resource`,
      [
        {
          id: 'AwsSolutions-IAM4',
          reason: 'AWS Custom resource provider framework-role created by cdk.',
        },
      ],
    );

    // AwsSolutions-IAM5: The IAM entity contains wildcard permissions
    NagSuppressions.addResourceSuppressionsByPath(
      stack,
      `${stack.stackName}/${id}/${providerId}/framework-onEvent/ServiceRole/DefaultPolicy/Resource`,
      [
        {
          id: 'AwsSolutions-IAM5',
          reason: 'Allows only specific policy.',
        },
      ],
    );
  }
}
