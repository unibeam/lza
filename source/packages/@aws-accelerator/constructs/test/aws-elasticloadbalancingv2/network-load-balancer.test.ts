/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { NetworkLoadBalancer } from '../../lib/aws-elasticloadbalancingv2/network-load-balancer';
import { snapShotTest } from '../snapshot-test';
import { describe } from '@jest/globals';

const testNamePrefix = 'Construct(NetworkLoadBalancer): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();

new NetworkLoadBalancer(stack, 'Test', {
  name: 'Test',
  ssmPrefix: '/accelerator',
  appName: 'appA',
  vpcName: 'vpcA',
  subnets: ['subnet-test123', 'subnet-test456'],
  deletionProtection: true,
  scheme: 'internal',
  crossZoneLoadBalancing: true,
  accessLogsBucket: 'test-bucket',
  listeners: [
    {
      name: 'string',
      certificate: 'fully-qualified-arn-acm',
      port: 80,
      protocol: 'HTTP',
      alpnPolicy: 'TLS_V1_2_2018',
      sslPolicy: 'ELBSecurityPolicy-2016-08',
      targetGroup: '${ACCEL_LOOKUP::TargetGroup:target-group-test}',
    },
    {
      name: 'string1',
      certificate: 'fully-qualified-arn-acm',
      port: 81,
      protocol: 'HTTP',
      alpnPolicy: 'TLS_V1_2_2018',
      sslPolicy: 'ELBSecurityPolicy-2016-08',
      targetGroup: 'target-group-test1',
    },
  ],
});

/**
 * NLB construct test
 */
describe('NetworkLoadBalancer', () => {
  snapShotTest(testNamePrefix, stack);
});
