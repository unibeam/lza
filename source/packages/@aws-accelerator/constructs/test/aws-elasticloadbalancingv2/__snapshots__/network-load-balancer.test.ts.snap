// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkLoadBalancer Construct(NetworkLoadBalancer):  Snapshot Test 1`] = `
{
  "Parameters": {
    "SsmParameterValueacceleratorapplicationtargetGroupappAvpcAtargetgrouptest1arnC96584B6F00A464EAD1953AFF4B05118Parameter": {
      "Default": "/accelerator/application/targetGroup/appA/vpcA/target-group-test1/arn",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "SsmParameterValueacceleratorapplicationtargetGroupappAvpcAtargetgrouptestarnC96584B6F00A464EAD1953AFF4B05118Parameter": {
      "Default": "/accelerator/application/targetGroup/appA/vpcA/target-group-test/arn",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Test7BFAF513": {
      "Properties": {
        "LoadBalancerAttributes": [
          {
            "Key": "deletion_protection.enabled",
            "Value": "true",
          },
          {
            "Key": "load_balancing.cross_zone.enabled",
            "Value": "true",
          },
          {
            "Key": "access_logs.s3.enabled",
            "Value": "true",
          },
          {
            "Key": "access_logs.s3.bucket",
            "Value": "test-bucket",
          },
          {
            "Key": "access_logs.s3.prefix",
            "Value": {
              "Fn::Join": [
                "",
                [
                  {
                    "Ref": "AWS::AccountId",
                  },
                  "/",
                  {
                    "Ref": "AWS::Region",
                  },
                  "/Test",
                ],
              ],
            },
          },
        ],
        "Name": "Test",
        "Scheme": "internal",
        "Subnets": [
          "subnet-test123",
          "subnet-test456",
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "Test",
          },
        ],
        "Type": "network",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "TestListenerstring15C484455": {
      "Properties": {
        "AlpnPolicy": [
          "TLS_V1_2_2018",
        ],
        "Certificates": [
          {
            "CertificateArn": "fully-qualified-arn-acm",
          },
        ],
        "DefaultActions": [
          {
            "ForwardConfig": {
              "TargetGroups": [
                {
                  "TargetGroupArn": {
                    "Ref": "SsmParameterValueacceleratorapplicationtargetGroupappAvpcAtargetgrouptest1arnC96584B6F00A464EAD1953AFF4B05118Parameter",
                  },
                },
              ],
            },
            "TargetGroupArn": {
              "Ref": "SsmParameterValueacceleratorapplicationtargetGroupappAvpcAtargetgrouptest1arnC96584B6F00A464EAD1953AFF4B05118Parameter",
            },
            "Type": "forward",
          },
        ],
        "LoadBalancerArn": {
          "Ref": "Test7BFAF513",
        },
        "Port": 81,
        "Protocol": "HTTP",
        "SslPolicy": "ELBSecurityPolicy-2016-08",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
    "TestListenerstring7A65A381": {
      "Properties": {
        "AlpnPolicy": [
          "TLS_V1_2_2018",
        ],
        "Certificates": [
          {
            "CertificateArn": "fully-qualified-arn-acm",
          },
        ],
        "DefaultActions": [
          {
            "ForwardConfig": {
              "TargetGroups": [
                {
                  "TargetGroupArn": {
                    "Ref": "SsmParameterValueacceleratorapplicationtargetGroupappAvpcAtargetgrouptestarnC96584B6F00A464EAD1953AFF4B05118Parameter",
                  },
                },
              ],
            },
            "TargetGroupArn": {
              "Ref": "SsmParameterValueacceleratorapplicationtargetGroupappAvpcAtargetgrouptestarnC96584B6F00A464EAD1953AFF4B05118Parameter",
            },
            "Type": "forward",
          },
        ],
        "LoadBalancerArn": {
          "Ref": "Test7BFAF513",
        },
        "Port": 80,
        "Protocol": "HTTP",
        "SslPolicy": "ELBSecurityPolicy-2016-08",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
  },
}
`;
